hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,123 @@
+3. Describe how this new state would be serialized and kept after restarts
+4. Define Readable and Writable stores for accessing and manipulating the pending airdrops state
+
+## Architecture
+
+The pending airdrop state would be part of the token state, since the airdrops are tightly connected to tokens and we can keep them at the same place. The pending nature of this state can be described as keeping incoming airdrops, without directly updating the sender and the receiver balances as in a regular transfer. This is needed to avoid transferring spam tokens and the receiver to retain full control of which airdrop to claim or ignore. Canceling an airdrop could be performed only from sender's side.
+
+New airdrop specific operations are also introduced - those are `TokenAirdrop`, `TokenClaimAirdrop` and `TokenCancelAirdrop`.
+
+Details for all of them would be covered in their own design docs, but here is a general overview of how they would interact with the pending state:
+- `TokenAirdrop` would add a new airdrop to the pending state
+- `TokenClaimAirdrop` would have a handle logic, which would take and remove a pending airdrop from this state, create a synthetic `CryptoTransfer` and execute it
+- `TokenCancelAirdrop` operation would remove an existing airdrop from the state
+
+### Structure
+
+We should define the structure of pending changes that would be kept in the `InitialModServiceTokenSchema` state, so we would reuse the token’s serialized state and add pending airdrops records in it. We could use the type defined in the protobuf records for the airdrop information itself. We should use the protobuf types since they are used for state serialization. So, the structure could look like this:
+
+`StateDefinition.*onDisk*(*AIRDROPS_KEY*, PendingAirdropId.PROTOBUF, PendingAirdropValue.PROTOBUF, MAX__AIRDROPS);`
+
+We can have a case, where we have multiple airdrops for the same fungible token. In this case the  `PendingAirdropValue` would be updated with the new aggregated value.
+
+### Stores
+
+We need to define a special `WritableStore` for the airdrops, so that we can manipulate the content in them. This would be the `WritableAirdropStore`.
+
+It would extend `ReadableAirdropStoreImpl` that implements the following interface:
+
+```java
+public interface ReadableAirdropStore {
+	PendingAirdropValue getFungibleAirdropAmount(@NonNull final PendingAirdropId airdropId);
+	long sizeOfState();
+	void warm(@NonNull final PendingAirdropId airdropId);
+    List<PendingAirdropId> getPendingAirdropIdsBySender(final AccountId sender);
+}
+```
+
+The `ReadableAirdropStoreImpl` would keep an instance of `ReadableKVState<PendingAirdropId, PendingAirdropValue> airdropsState`.
+
+An additional method `getPendingAirdropIdsBySender` would be added to the `ReadableAirdropStore` to get all `PendingAirdropIds` for a specific sender. This would be done via filtering all `PendingAirdropIds` kept in state by the `sender` that is passed as a parameter.
```

Comment:
This will not work. We cannot filter all entries in the state. Please check how this is done in similar cases (e.g., for token associations).

Reply:
I changed the design and aligned it with the `TokenRelationStore`. Instead of filtering inside of state, now the design doc reflects getting the airdrop keys from state and iterate over them from the outside and perform business logic there.